File: src/main/nashorn/net/coderextreme/data/HAnimModelsHandsFeet.js
load('X3Dautoclass.js');
var ConfigurationProperties = Packages.org.web3d.x3d.jsail.ConfigurationProperties;
ConfigurationProperties.showDefaultAttributes = false;
ConfigurationProperties.xsltEngine = ConfigurationProperties.XSLT_ENGINE_NATIVE_JAVA;
ConfigurationProperties.deleteIntermediateFiles = false;
ConfigurationProperties.setStripTrailingZeroes(true);
      var X3D0 =  new X3D().setProfile("Immersive").setVersion("4.0")
      .setHead(new head()
        .addComponent(new component().setName("HAnim").setLevel(1))
        .addMeta(new meta().setName("title").setContent("HAnimModelsHandsFeet.x3d"))
        .addMeta(new meta().setName("description").setContent("Left and right hands and feet, using high-fidelity definitions for HAnim version 2.0"))
        .addMeta(new meta().setName("creator").setContent("YOO Kwan Hee and Don Brutzman"))
        .addMeta(new meta().setName("created").setContent("8 February 2015"))
        .addMeta(new meta().setName("modified").setContent("8 March 2021"))
        .addMeta(new meta().setName("warning").setContent("not yet to scale"))
        .addMeta(new meta().setName("warning").setContent("TODO will X3D HAnim component add a new level to support LOA-4 functionality?"))
        .addMeta(new meta().setName("TODO").setContent("how to have HAnimHumanoid root with Inline IMPORT/EXPORT of limbs?"))
        .addMeta(new meta().setName("Image").setContent("HAnimModelsHandsFeet.png"))
        .addMeta(new meta().setName("Image").setContent("HAnimModelsHandsFeetWithFour1mGrids.png"))
        .addMeta(new meta().setName("reference").setContent("https://www.web3d.org/working-groups/humanoid-animation-HAnim"))
        .addMeta(new meta().setName("reference").setContent("https://www.web3d.org/documents/specifications/19774/V2.0"))
        .addMeta(new meta().setName("reference").setContent("https://www.web3d.org/specifications/X3Dv4Draft/ISO-IEC19775-1v4-WD2/Part01/components/hanim.html"))
        .addMeta(new meta().setName("subject").setContent("X3D HAnim humanoid animation"))
        .addMeta(new meta().setName("identifier").setContent("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelsHandsFeet.x3d"))
        .addMeta(new meta().setName("generator").setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"))
        .addMeta(new meta().setName("license").setContent("../license.html")))
      .setScene(new Scene()
        .addChild(new WorldInfo().setTitle("HAnimModelsHandsFeet.x3d"))
        .addChild(new Viewpoint().setDescription("Hands and feet 10m"))
        .addChild(new Viewpoint().setDescription("Hands and feet 1.7m").setPosition(Java.to([0,0,1.7], Java.type("float[]"))))
        .addChild(new Transform().setTranslation(Java.to([-1,1,0], Java.type("float[]")))
          .addChild(new Inline().setUrl(Java.to(["HAnimModelHandLeft.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandLeft.x3d","HAnimModelHandLeft.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandLeft.wrl"], Java.type("java.lang.String[]"))))
          .addComments(new CommentsBlock("Grid overlay authoring hint: first adjust grid scale to convenient large size, then adjust overall scale for your scene model"))
          .addChild(new Transform().setDEF("GridXY_20x20Fixed_AdjustScale").setScale(Java.to([0.1,0.1,0.1], Java.type("float[]")))
            .addChild(new Inline().setDEF("GridXY_20x20Fixed").setUrl(Java.to(["GridXY_20x20Fixed.x3d","../../Savage/Tools/Authoring/GridXY_20x20Fixed.x3d","https://savage.nps.edu/Savage/Tools/Authoring/GridXY_20x20Fixed.x3d","GridXY_20x20Fixed.wrl","../../Savage/Tools/Authoring/GridXY_20x20Fixed.wrl","https://savage.nps.edu/Savage/Tools/Authoring/GridXY_20x20Fixed.wrl"], Java.type("java.lang.String[]"))))))
        .addChild(new Transform().setTranslation(Java.to([1,1,0], Java.type("float[]")))
          .addChild(new Inline().setUrl(Java.to(["HAnimModelHandRight.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandRight.x3d","HAnimModelHandRight.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandRight.wrl"], Java.type("java.lang.String[]"))))
          .addChild(new Transform().setUSE("GridXY_20x20Fixed_AdjustScale")))
        .addChild(new Transform().setTranslation(Java.to([-1,-1,0], Java.type("float[]")))
          .addComments(new CommentsBlock("rotation='0 0 1 3.141593'"))
          .addChild(new Inline().setUrl(Java.to(["HAnimModelFootLeft.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootLeft.x3d","HAnimModelFootLeft.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootLeft.wrl"], Java.type("java.lang.String[]"))))
          .addChild(new Transform().setUSE("GridXY_20x20Fixed_AdjustScale")))
        .addChild(new Transform().setTranslation(Java.to([1,-1,0], Java.type("float[]")))
          .addComments(new CommentsBlock("rotation='0 0 1 3.141593'"))
          .addChild(new Inline().setUrl(Java.to(["HAnimModelFootRight.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootRight.x3d","HAnimModelFootRight.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootRight.wrl"], Java.type("java.lang.String[]"))))
          .addChild(new Transform().setUSE("GridXY_20x20Fixed_AdjustScale"))))      ;
    X3D0.toFileX3D("../data/HAnimModelsHandsFeet.new.x3d");

File: src/main/quarter/net/coderextreme/data/HAnimModelsHandsFeet.cpp

#include <Inventor/nodes/SoCone.h>
#include <Inventor/nodes/SoBaseColor.h>
#include <Inventor/nodes/SoSeparator.h>
#include <Inventor/nodes/SoShape.h>
#include <Inventor/nodes/SoTransform.h>
#include <Inventor/nodes/SoSphere.h>
#include <Inventor/nodes/SoMaterial.h>
#include <Inventor/misc/SoProtoInstance.h>
#include <Inventor/VRMLnodes/SoVRMLAppearance.h>
#include <Quarter/Quarter.h>
#include <Quarter/QuarterWidget.h>
#include <QApplication>
#include <QMainWindow>

using namespace SIM::Coin3D::Quarter;
int main(int argc, char ** argv) 
{

  QApplication app(argc, argv);
  Quarter::init();

  SoSeparator * root = new SoSeparator;
  root->ref();

  SoBaseColor * col = new SoBaseColor;
  col->rgb = SbColor(1, 1, 0);
  root->addChild(col);

  root->addChild(new SoCone);

  QMainWindow * mainwin = new QMainWindow();

  QuarterWidget * viewer = new QuarterWidget(mainwin);
  viewer->setNavigationModeFile();
  mainwin->setCentralWidget(viewer);
  viewer->setSceneGraph(root);

  mainwin->show();
  app.exec();
  root->unref();
  delete viewer;

  Quarter::clean();
SoSceneManager* SoSceneManager0 = new SoSceneManager();
SoSceneManager0->setProfile(QString("Immersive"));
SoSceneManager0->setVersion(QString("4.0"));
Sohead* Sohead1 = new Sohead();
Socomponent* Socomponent2 = new Socomponent();
Socomponent2->setName(QString("HAnim"));
Socomponent2->setLevel(1);
Sohead1->addChild(*Socomponent2);

Someta* Someta3 = new Someta();
Someta3->setName(QString("title"));
Someta3->setContent(QString("HAnimModelsHandsFeet.x3d"));
Sohead1->addMeta(*Someta3);

Someta* Someta4 = new Someta();
Someta4->setName(QString("description"));
Someta4->setContent(QString("Left and right hands and feet, using high-fidelity definitions for HAnim version 2.0"));
Sohead1->addMeta(*Someta4);

Someta* Someta5 = new Someta();
Someta5->setName(QString("creator"));
Someta5->setContent(QString("Kwan Hee YOO and Don Brutzman"));
Sohead1->addMeta(*Someta5);

Someta* Someta6 = new Someta();
Someta6->setName(QString("created"));
Someta6->setContent(QString("8 February 2015"));
Sohead1->addMeta(*Someta6);

Someta* Someta7 = new Someta();
Someta7->setName(QString("modified"));
Someta7->setContent(QString("Mon, 15 Sep 2025 05:20:09 GMT"));
Sohead1->addMeta(*Someta7);

Someta* Someta8 = new Someta();
Someta8->setName(QString("warning"));
Someta8->setContent(QString("not yet to scale"));
Sohead1->addMeta(*Someta8);

Someta* Someta9 = new Someta();
Someta9->setName(QString("warning"));
Someta9->setContent(QString("TODO will X3D HAnim component add a new level to support LOA-4 functionality?"));
Sohead1->addMeta(*Someta9);

Someta* Someta10 = new Someta();
Someta10->setName(QString("TODO"));
Someta10->setContent(QString("how to have HAnimHumanoid root with Inline IMPORT/EXPORT of limbs?"));
Sohead1->addMeta(*Someta10);

Someta* Someta11 = new Someta();
Someta11->setName(QString("Image"));
Someta11->setContent(QString("HAnimModelsHandsFeet.png"));
Sohead1->addMeta(*Someta11);

Someta* Someta12 = new Someta();
Someta12->setName(QString("Image"));
Someta12->setContent(QString("HAnimModelsHandsFeetWithFour1mGrids.png"));
Sohead1->addMeta(*Someta12);

Someta* Someta13 = new Someta();
Someta13->setName(QString("reference"));
Someta13->setContent(QString("https://www.web3d.org/working-groups/humanoid-animation-HAnim"));
Sohead1->addMeta(*Someta13);

Someta* Someta14 = new Someta();
Someta14->setName(QString("reference"));
Someta14->setContent(QString("https://www.web3d.org/documents/specifications/19774/V2.0"));
Sohead1->addMeta(*Someta14);

Someta* Someta15 = new Someta();
Someta15->setName(QString("reference"));
Someta15->setContent(QString("https://www.web3d.org/specifications/X3Dv4/ISO-IEC19775-1v4-IS/Part01/components/hanim.html"));
Sohead1->addMeta(*Someta15);

Someta* Someta16 = new Someta();
Someta16->setName(QString("subject"));
Someta16->setContent(QString("X3D HAnim humanoid animation"));
Sohead1->addMeta(*Someta16);

Someta* Someta17 = new Someta();
Someta17->setName(QString("identifier"));
Someta17->setContent(QString("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelsHandsFeet.x3d"));
Sohead1->addMeta(*Someta17);

SoSceneManager0->setHead(*Sohead1);

SoNode* SoNode18 = new SoNode();
SoWorldInfo* SoWorldInfo19 = new SoWorldInfo();
SoWorldInfo19->setTitle(QString("HAnimModelsHandsFeet.x3d"));
SoNode18->addChild(*SoWorldInfo19);

SoViewpoint* SoViewpoint20 = new SoViewpoint();
SoViewpoint20->setDescription(QString("Hands and feet 10m"));
SoNode18->addChild(*SoViewpoint20);

SoViewpoint* SoViewpoint21 = new SoViewpoint();
SoViewpoint21->setDescription(QString("Hands and feet 1.7m"));
SoViewpoint21->setPosition(new float[]{0.0,0.0,1.7});
SoNode18->addChild(*SoViewpoint21);

SoTransform* SoTransform22 = new SoTransform();
SoTransform22->setTranslation(new float[]{-1.0,1.0,0.0});
SoInline* SoInline23 = new SoInline();
SoInline23->setUrl(new QString[]{QString("HAnimModelHandLeft.x3d"), QString("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandLeft.x3d"), QString("HAnimModelHandLeft.wrl"), QString("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandLeft.wrl")}, 4);
SoTransform22->addChild(*SoInline23);

SoTransform* SoTransform24 = new SoTransform();
SoTransform24->setDEF(QString("GridXY_20x20Fixed_AdjustScale"));
SoTransform24->setScale(new float[]{0.1,0.1,0.1});
SoInline* SoInline25 = new SoInline();
SoInline25->setDEF(QString("GridXY_20x20Fixed"));
SoInline25->setUrl(new QString[]{QString("../../Savage/Tools/Authoring/GridXY_20x20Fixed.x3d"), QString("https://savage.nps.edu/Savage/Tools/Authoring/GridXY_20x20Fixed.x3d"), QString("../../Savage/Tools/Authoring/GridXY_20x20Fixed.wrl"), QString("https://savage.nps.edu/Savage/Tools/Authoring/GridXY_20x20Fixed.wrl")}, 4);
SoTransform24->addChild(*SoInline25);

SoTransform22->addChild(*SoTransform24);

SoNode18->addChild(*SoTransform22);

SoTransform* SoTransform26 = new SoTransform();
SoTransform26->setTranslation(new float[]{1.0,1.0,0.0});
SoInline* SoInline27 = new SoInline();
SoInline27->setUrl(new QString[]{QString("HAnimModelHandRight.x3d"), QString("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandRight.x3d"), QString("HAnimModelHandRight.wrl"), QString("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandRight.wrl")}, 4);
SoTransform26->addChild(*SoInline27);

SoTransform* SoTransform28 = new SoTransform();
SoTransform28->setUSE(QString("GridXY_20x20Fixed_AdjustScale"));
SoTransform26->addChild(*SoTransform28);

SoNode18->addChild(*SoTransform26);

SoTransform* SoTransform29 = new SoTransform();
SoTransform29->setTranslation(new float[]{-1.0,-1.0,0.0});
SoInline* SoInline30 = new SoInline();
SoInline30->setUrl(new QString[]{QString("HAnimModelFootLeft.x3d"), QString("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootLeft.x3d"), QString("HAnimModelFootLeft.wrl"), QString("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootLeft.wrl")}, 4);
SoTransform29->addChild(*SoInline30);

SoTransform* SoTransform31 = new SoTransform();
SoTransform31->setUSE(QString("GridXY_20x20Fixed_AdjustScale"));
SoTransform29->addChild(*SoTransform31);

SoNode18->addChild(*SoTransform29);

SoTransform* SoTransform32 = new SoTransform();
SoTransform32->setTranslation(new float[]{1.0,-1.0,0.0});
SoInline* SoInline33 = new SoInline();
SoInline33->setUrl(new QString[]{QString("HAnimModelFootRight.x3d"), QString("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootRight.x3d"), QString("HAnimModelFootRight.wrl"), QString("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootRight.wrl")}, 4);
SoTransform32->addChild(*SoInline33);

SoTransform* SoTransform34 = new SoTransform();
SoTransform34->setUSE(QString("GridXY_20x20Fixed_AdjustScale"));
SoTransform32->addChild(*SoTransform34);

SoNode18->addChild(*SoTransform32);

SoSceneManager0->setSceneGraph(*SoNode18);

return 0;
}
